### PR TITLE
Fixing issue with double tapped part creation. Adding "smart add"

### PR DESCRIPTION
**Pull request by vinniefranco**
## What's new?
* Smart quantity increment/split for bundles when using product pages to "add" quantity to user selectable bundles
* Fix for #120 

## Smart increment/split rational.
With user selectable variants we run into a bit of weirdness...

**Example Setup**

**Product A:** 2 Variants - RED, BLUE
**Product B:** 2 Variants - Large, Medium
**Product C:** Bundles Product A & B.

**Current scenario:**

1. User adds C with RED, and Large.
2. User navigates back to product detail.
3. User adds C with BLUE, and Large.

**Cart shows:**
2 Product C with (2) Large, (1) BLUE, (1) RED

This doesn't really make sense from a bundling perspective, especially for decrements in quantity. What did you want to remove?

##### Scenario 1:

1. User adds C with RED, and Large.
2. User navigates back to product detail.
3. User adds C with BLUE, and Large.

**Cart shows:**

(1) Product C with (1) RED, (1) Large
(1) Product C with (1) BLUE, (1) Large

##### Scenario 2:

1. User adds C with RED, and Large.
2. User navigates back to product detail.
3. User adds C with BLUE, and Medium.

**Cart shows:**

(1) Product C with (1) RED, (1) Large
(1) Product C with (1) BLUE, (1) Medium

##### Scenario 3:

1. User adds C with RED, and Large.
2. User navigates back to product detail.
3. User adds C with RED, and Large.

**Cart shows:**

(2) Product C with (2) RED, (2) Large

### CODE REVIEW

File: app/models/spree/order_contents_decorator.rb [hunk 1]
```diff
@@ -1,5 +1,26 @@
 module Spree
   OrderContents.class_eval do
+
+    def add_to_line_item(variant, quantity, options = {})
+      line_item = grab_line_item_by_variant(variant, false, options)
+
+      if line_item && part_variants_match(line_item, variant, quantity, options)
+        line_item.quantity += quantity.to_i
+        line_item.currency = currency unless currency.nil?
```
Comment (sikachu) [inline app/models/spree/order_contents_decorator.rb]: YAGNI on the conditional. You'll assign a `nil` regardless?

Comment (sikachu) [inline app/models/spree/order_contents_decorator.rb]: Or, if you want to not overriding `line_item.currency` if `currency` is nil, maybe this is fine as-is.

File: app/models/spree/order_contents_decorator.rb [hunk 2]
```diff
@@ -1,5 +1,26 @@
...
+      else
+        opts = { currency: order.currency }.merge ActionController::Parameters.new(options).
+                                            permit(PermittedAttributes.line_item_attributes)
+        line_item = order.line_items.new(quantity: quantity,
+                                          variant: variant,
+                                          options: opts)
+      end
+
+      line_item.target_shipment = options[:shipment] if options.has_key? :shipment
```
Comment (sikachu) [inline app/models/spree/order_contents_decorator.rb]: Somewhat same as above. Does `line_item.target_shipment` initially holds a value but can be overridden?

File: app/models/spree/order_contents_decorator.rb [hunk 3]
```diff
@@ -10,17 +31,34 @@ def add_to_line_item_with_parts(variant, quantity, options = {})
         )
       end
     end
+
     alias_method_chain :add_to_line_item, :parts
 
     private
 
+    def part_variants_match(line_item, variant, quantity, options)
```
Comment (sikachu) [inline app/models/spree/order_contents_decorator.rb]: Considered rename this method to `part_variants_match?` since it returns boolean.

File: app/models/spree/order_contents_decorator.rb [hunk 4]
```diff
@@ -10,17 +31,34 @@ def add_to_line_item_with_parts(variant, quantity, options = {})
...
       parts.each do |part|
-        line_item.part_line_items.create!(
+        part_line_item = line_item.part_line_items.find_or_initialize_by(
           line_item: line_item,
-          variant_id: variant_id_for(part, selected_variants),
-          quantity: part.count
+          variant_id: variant_id_for(part, selected_variants)
         )
+
+        part_line_item.update_attributes(quantity: part.count)
```
Comment (sikachu) [inline app/models/spree/order_contents_decorator.rb]: Do we have something to handle this line if it return `false`? Should it be `update_attributes!`?

File: app/models/spree/order_contents_decorator.rb (hunk 1 above)
Comment (vinniefranco) [inline app/models/spree/order_contents_decorator.rb]: That's pulled straight from spree haha. I just added && part_variants_match

File: spec/features/adding_items_to_cart_spec.rb [hunk 5]
```diff
@@ -155,6 +155,130 @@
...
+      end
+      let(:shirt) do
+        shirt, _shirts_by_size = create_bundle_product_with_options(
+          name: "Shirt",
+          sku: "SHIRT",
+          option_type: "Size",
+          option_values: ["Large", "XL"]
+        )
+
+        shirt
```
Comment (sikachu) [inline spec/features/adding_items_to_cart_spec.rb]: Does it make sense to create another helper method that calls `create_bundle_product_with_options` internally but return only the product?

File: spec/features/adding_items_to_cart_spec.rb [hunk 6]
```diff
@@ -155,6 +155,130 @@
...
+          select "Color: Red", from: "options_selected_variants_3"
+          click_button "add-to-cart-button"
+
+          visit spree.product_path(bundle)
+
+          select "Size: XL", from: "options_selected_variants_2"
+          select "Color: Blue", from: "options_selected_variants_3"
+          click_button "add-to-cart-button"
+
+          within all("#cart-detail .line-item")[0] do
```
Comment (sikachu) [inline spec/features/adding_items_to_cart_spec.rb]: Minor, but I think this `all("#cart-detail .line-item")[0]` and on line 236 can be refactored as a well named method for better readability.

File: spec/features/adding_items_to_cart_spec.rb [hunk 7]
```diff
@@ -155,6 +155,130 @@
...
+          variant_selection_deferred: true
+        )
+      end
+
+      context "and the user selects differing variants from the existing line item" do
+        specify "the cart contains incremented variants when listing bundle items" do
+
+          visit spree.product_path(bundle)
+
+          select "Size: Large", from: "options_selected_variants_2"
```
Comment (sikachu) [inline spec/features/adding_items_to_cart_spec.rb]: I feel like these three lines can be refactored as a well named method as well. Maybe `add_item_to_cart(size: "Large", color: "Red")`. That way, if the field ever change later we don't have to change 8 places.

File: spec/features/adding_items_to_cart_spec.rb (hunk 5 above)
Comment (vinniefranco) [inline spec/features/adding_items_to_cart_spec.rb]: I think so. I was debating doing so, but I think you're right. I'll get on
that

File: app/models/spree/order_contents_decorator.rb (hunk 3 above)
Comment (vinniefranco) [inline app/models/spree/order_contents_decorator.rb]: Excellent

File: app/models/spree/order_contents_decorator.rb (hunk 1 above)
Comment (sikachu) [inline app/models/spree/order_contents_decorator.rb]: I see. In that case I think it's okay to leave it as-is. I don't want us to take a risk in case the code changes in the future.

**Comment by vinniefranco** [comment]
@futhr I need another set of eyes on this. Once this is in I'll get cracking on the upgrade to 3.x

**Comment by vinniefranco** [comment]
@joneslee85 any time to review?

File: app/models/spree/order_contents_decorator.rb [hunk 8]
```diff
@@ -1,5 +1,26 @@
 module Spree
   OrderContents.class_eval do
+
+    def add_to_line_item(variant, quantity, options = {})
+      line_item = grab_line_item_by_variant(variant, false, options)
+
+      if line_item && part_variants_match?(line_item, variant, quantity, options)
+        line_item.quantity += quantity.to_i
+        line_item.currency = currency unless currency.nil?
```
Comment (elle) [inline app/models/spree/order_contents_decorator.rb]: What are your thoughts on double negatives? I usually prefer to avoid them

File: app/models/spree/order_contents_decorator.rb [hunk 9]
```diff
@@ -1,5 +1,26 @@
...
+      else
+        opts = { currency: order.currency }.merge ActionController::Parameters.new(options).
+                                            permit(PermittedAttributes.line_item_attributes)
+        line_item = order.line_items.new(quantity: quantity,
+                                          variant: variant,
+                                          options: opts)
+      end
+
+      line_item.target_shipment = options[:shipment] if options.has_key? :shipment
```
Comment (elle) [inline app/models/spree/order_contents_decorator.rb]: Suggestion to rewrite as:
`line_item.target_shipment = options.fetch(:shipment, nil)`

File: app/models/spree/order_contents_decorator.rb [hunk 10]
```diff
@@ -1,5 +1,26 @@
...
   OrderContents.class_eval do
+
+    def add_to_line_item(variant, quantity, options = {})
+      line_item = grab_line_item_by_variant(variant, false, options)
+
+      if line_item && part_variants_match?(line_item, variant, quantity, options)
+        line_item.quantity += quantity.to_i
+        line_item.currency = currency unless currency.nil?
+      else
+        opts = { currency: order.currency }.merge ActionController::Parameters.new(options).
```
Comment (elle) [inline app/models/spree/order_contents_decorator.rb]: Why do we go through `ActionController::Parameters.new` here?
Would it work with `{ currency: order.currency }.merge(options)`?

File: app/models/spree/order_contents_decorator.rb (hunk 8 above)
Comment (elle) [inline app/models/spree/order_contents_decorator.rb]: Also, where is currency defined in this class? or should this be `order.currency` as in line 11?

File: spec/features/adding_items_to_cart_spec.rb [hunk 11]
```diff
@@ -155,6 +153,117 @@
...
+  def add_item_to_cart(args)
+    visit spree.product_path(bundle)
+
+    select "Size: #{args[:size]}", from: "options_selected_variants_2"
+    select "Color: #{args[:color]}", from: "options_selected_variants_3"
+    click_button "add-to-cart-button"
+  end
+
+  def bundled_product_from_options(args)
+    args[:sku] ||= args[:name].parameterize.upcase
```
Comment (elle) [inline spec/features/adding_items_to_cart_spec.rb]: Does this need to be memoized?

File: spec/features/adding_items_to_cart_spec.rb [hunk 12]
```diff
@@ -155,6 +153,117 @@
...
+          variant_selection_deferred: true
+        )
+      end
+
+      context "and the user selects differing variants from the existing line item" do
+        specify "the cart contains incremented variants when listing bundle items" do
+          add_item_to_cart(size: "Large", color: "Red")
+          add_item_to_cart(size: "XL", color: "Blue")
+
+          within all("#cart-detail .line-item")[0] do
```
Comment (elle) [inline spec/features/adding_items_to_cart_spec.rb]: Just checking that the line items markup does not have any other unique identifiers to be used as clearer selectors for Capybara...

File: spec/features/adding_items_to_cart_spec.rb (hunk 11 above)
Comment (vinniefranco) [inline spec/features/adding_items_to_cart_spec.rb]: It's just to provide a default, if you provide a sku nothing gets set, if you don't name becomes sku

File: app/models/spree/order_contents_decorator.rb (hunk 10 above)
Comment (vinniefranco) [inline app/models/spree/order_contents_decorator.rb]: It probably would, but this is core. Not really wanting to deviate from what they have: https://github.com/spree/spree/blob/master/core/app/models/spree/order_contents.rb#L80

File: spec/features/adding_items_to_cart_spec.rb (hunk 12 above)
Comment (vinniefranco) [inline spec/features/adding_items_to_cart_spec.rb]: Doesn't appear to be any...
screen shot 2015-09-03 at 1 53 08 pm

File: app/models/spree/order_contents_decorator.rb (hunk 8 above)
Comment (vinniefranco) [inline app/models/spree/order_contents_decorator.rb]: It's core... Should we spend the time to refactor core in this PR or should it be a separate PR to core itself?
